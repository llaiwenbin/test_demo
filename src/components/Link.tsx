// Copyright 2004-present Facebook. All Rights Reserved.

const useState = require('react').useState;

const STATUS = {
  HOVERED: 'hovered',
  NORMAL: 'normal',
};

function Link({ page, children }) {
  const [status, setStatus] = useState(STATUS.NORMAL);

  const onMouseEnter = () => {
    setStatus(STATUS.HOVERED);
  };

  const onMouseLeave = () => {
    setStatus(STATUS.NORMAL);
  };

  return (
    <a
      className={status}
      href={page || '#'}
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
    >
      {children}
    </a>
  );
}
export default Link