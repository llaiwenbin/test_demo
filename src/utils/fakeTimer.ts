const sleep = async (time: number, result: string): Promise<string> => {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve(result);
    }, time);
  });
};


const asyncSleep = async (time: number, fn: () => void): Promise<void> => {
  setTimeout(() => {
    Promise.resolve().then(() => {
      fn();
    });
  }, time);
};


export { sleep, asyncSleep };