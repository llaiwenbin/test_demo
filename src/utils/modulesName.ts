const foo = () => { };
const def = () => { };



export { foo };
export default def