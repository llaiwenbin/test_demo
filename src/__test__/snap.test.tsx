import renderer from 'react-test-renderer';
import Link from '../components/Link';
import { asyncSleep } from '../utils/FakeTimer';
// 快照：一般对稳定不修改的方法或者组件进行快照，这样可以节约单元测试时间可以保证dom结构不改变
// https://jestjs.io/docs/snapshot-testing
describe('快照', () => {
  // 对方法进行快照 -- 还是推荐用单测的断言吧
  test('renders correctly', () => {
    expect(asyncSleep.toString()).toMatchSnapshot();
  })

  // 只对最终渲染效果进行快照，渲染效果一般和输入相关，如果输入改变也可能导致快照不通过
  test('对组件进行快照', () => {
    const link = renderer
      .create(<Link page="http://www.facebook.com">Facebook</Link>)
      .toJSON();
    expect(link).toMatchSnapshot()

    // 修改输入快照失败
    // const link = renderer
    //   .create(<Link page="123">Facebook</Link>)
    //   .toJSON();
    // expect(link).toMatchSnapshot()
  })
  
  // 通过提供的匹配器进行快照
  it('对变量对象进行快照', () => {
    const user = {
      createdAt: new Date(),
      id: Math.floor(Math.random() * 20),
      name: 'LeBron James',
    };

    expect(user).toMatchSnapshot({
      createdAt: expect.any(Date),
      id: expect.any(Number),
    });
  });

});
