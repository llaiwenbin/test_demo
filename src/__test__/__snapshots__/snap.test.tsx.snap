// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`快照 renders correctly 1`] = `
"async (time, fn) => {
  /* istanbul ignore next */
  cov_1tbr90e05i().f[3]++;
  cov_1tbr90e05i().s[5]++;
  setTimeout(() => {
    /* istanbul ignore next */
    cov_1tbr90e05i().f[4]++;
    cov_1tbr90e05i().s[6]++;
    Promise.resolve().then(() => {
      /* istanbul ignore next */
      cov_1tbr90e05i().f[5]++;
      cov_1tbr90e05i().s[7]++;
      fn();
    });
  }, time);
}"
`;

exports[`快照 will check the matchers and pass 1`] = `
{
  "createdAt": Any<Date>,
  "id": Any<Number>,
  "name": "LeBron James",
}
`;

exports[`快照 对变量对象进行快照 1`] = `
{
  "createdAt": Any<Date>,
  "id": Any<Number>,
  "name": "LeBron James",
}
`;

exports[`快照 对组件进行快照 1`] = `
<a
  className="normal"
  href="http://www.facebook.com"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  Facebook
</a>
`;
